fplll/pruner/pruner.h

```diff
@@ -290,37 +386,135 @@ template <class FT> class Pruner
+    }
+    else if (metric == PRUNER_METRIC_EXPECTED_SOLUTIONS)
+    {
+      // check if the target is reasonable; otherwise set it to 0.99.
+      // note the target is allowed to be larger than 1 in this case.
+      // Also the target may affect the cost in target_function().
+      if (this->target < 0.0)
+      {
+        this->target = 0.99;
+      }
+    }
+    else
+    {
+      throw std::invalid_argument("Pruner was set to an unknown metric");
+    }
     load_basis_shapes(gso_rs);
   }
 
   /**
-     Run the optimization process, successively using the algorithm activated
+     @brief Main interface to optimize pruning coefficients
+
+     Main interface to optimize pruning coefficients.
+     It will in turn invoke either of the two functions:
+      (1)   optimize_coefficients_cost_vary_prob() or
+      (2)   optimize_coefficients_cost_fixed_prob()
+     depending on the input "target".
 
-     @brief run the optimization process
+     If the target is negative (e.g. -1), it calls function (1)
+     such that goal is to optimize the
+     single_enum_cost(pr) * trials + preproc_cost * (trials - 1.0).
+
+     If the target is > 0, it calls function (2) such that the
+     goal is to optimize the single_enum_cost(pr) while fixing
+     the succ. probability == target.
   */
   void optimize_coefficients(/*io*/ vector<double> &pr);
 
-  /** @brief Compute the cost of a single enumeration */
+  /**
+     @brief main interface to optimize the pruning coefficients with
+     respect to the overall enumeraiton time.
+
+     Main interface to optimize the overall enumeraiton time where the
+     target function is:
+
+     single_enum_cost(pr) * trials + preproc_cost * (trials - 1.0);
+  */
+  void optimize_coefficients_cost_vary_prob(/*io*/ vector<double> &pr);
+
+  /**
+     @brief main interface to optimize the pruning coefficients with
+     repeect to the single enumeraiton time while fixing the succ. prob
+     or expected number of solutions.
+
+     Main interface to optimize the single enumeration time with the
+     constraint such that the succ. prob (or expected solutions) is
+     fixed (and given) from input to the Pruner constructor.
+  */
+  void optimize_coefficients_cost_fixed_prob(/*io*/ vector<double> &pr);
 
+  /**
+     @brief run the optimization process using 'even' coefficients. Note
+     the optimization only applies to the pruning coefficients indexed
+     by (0, 2, 4, ... n). It thus uses half of the coefficients.
+
+     Run the optimization process, successively using the algorithm activated
+     using using half coefficients: the input pr has length n; but only the
+     even indices in the vector will be used in the optimization.
+     In the end, we have pr_i = pr_{i+1}.
+     Note it only optimize the overall enumeraiton time where the
+     target function is:
+     single_enum_cost(pr) * trials + preproc_cost * (trials - 1.0);
+  */
+  void optimize_coefficients_evec(/*io*/ vector<double> &pr);
```

Comment:
Maybe it is useful to explain the hierarchy of functions here: `optimize_coefficients` calls `optimize_coefficients_vary_prob` which calls `optimize_coefficients_evec`. Perhaps the latter should also be renamed to `optimize_coefficients_vary_prob_evec`?

Reply:
I added some explanation on the hierarchy of functions. The function optimize_coefficients_evec() is used in both optimize_coefficients_vary_prob() and optimize_coefficients_cost_fixed_prob() as a common function.